proxmoxtf/resource/vm.go

```diff
@@ -3997,7 +3996,7 @@ func vmReadCustom(
 		pciMap[pi] = pci
 	}
 
-	if len(currentPCIList) > 0 {
```

Comment:
Hmmmmm... this looks a bit suspicious -- it will write the to the state regardless if the plan has hostpci defined or not, which may cause some side-effects.
I'd like to spend some more time testing the change.

Reply:
Thanks for your feedback! It is very much possible that I misunderstood the code that I changed.
Anyway, during my tests, `currentPCIList` was empty which lead to the fix I provided. Does `schema.ResourceData` (that's where currrentPCIList is taken from) represent the Terraform inventory or the current state?

Reply:
To be clear, when i said "state" and "plan" i meant both TF state and TF plan, regardless what is in PVE.

In this context, `d schema.ResourceData` represent them both, when it is used in "Create" operation. `d.Get` returns an attribute from the currently applying plan, `d.Set` sets an attribute that then will be written to the state at the end of apply.

So with this change _I think_ is possible to have a situation when hostpci is not defined in the plan, but exists in the VM in PVE - for example, if VM is cloned form one that has hostpci. In this case TF will write in the state those extra attributes at the end. Then if you re-apply the exact same plan, TF will detect discrepancy between plan and state - the hostpci section is not in the plan but is in the state, and will try to remove hostpci.

Again, this is just a suspicious at the moment, after looking at the code. I haven't had a chance to test this case properly yet.

Reply:
I was able to reproduce the situation I was describing above.

Template:
```hcl
resource "proxmox_virtual_environment_vm" "ubuntu_vm" {
  name = "578-hostpci"
  node_name = "pve"
  vm_id     = 578

  hostpci {
    device = "hostpci0"
    id = "0000:01:00.0"
    pcie = true
  }

  started = false
}

resource "proxmox_virtual_environment_vm" "ubuntu_vm_cloned" {
  depends_on = [proxmox_virtual_environment_vm.ubuntu_vm]
  name = "579-hostpci-cloned"
  node_name = "pve"
  vm_id     = 579

  clone {
    vm_id = 578
  }

  started = false
}
```

After `terrafrom apply` the cloned VM got deployed with its `hostpci` device:
<img width="724" alt="Screenshot 2023-09-24 at 8 43 31 PM" src="https://github.com/bpg/terraform-provider-proxmox/assets/627562/85b3d946-97a7-40f6-bedf-7d9a4bc638d2">

But after the second apply, without any changes to the plan, the `hostpci` in the clone got removed:
```
...
 # proxmox_virtual_environment_vm.ubuntu_vm_cloned will be updated in-place
  ~ resource "proxmox_virtual_environment_vm" "ubuntu_vm_cloned" {
        id                      = "579"
      + mac_addresses           = (known after apply)
        name                    = "579-hostpci-cloned"
        # (22 unchanged attributes hidden)

      - hostpci {
          - device = "hostpci0" -> null
          - id     = "0000:01:00.0" -> null
          - pcie   = true -> null
          - rombar = false -> null
          - xvga   = false -> null
        }

        # (1 unchanged block hidden)
    }
...
```
<img width="773" alt="Screenshot 2023-09-24 at 8 44 33 PM" src="https://github.com/bpg/terraform-provider-proxmox/assets/627562/ae475d9f-ea60-4472-a889-8fa3690c8b6c">

Reply:
I've also tried to reproduce the original issue (as described in #577) on v0.32.0:

Template:
```hcl
resource "proxmox_virtual_environment_vm" "ubuntu_vm" {
  name = "578-hostpci"
  node_name = "pve"
  vm_id     = 578

#  hostpci {
#    device = "hostpci0"
#    id = "0000:01:00.0"
#    pcie = true
#  }

  started = false
}
```

Apply, then uncomment the `hostpci` block, and apply again: the `hostpci` devices got created 🤔
```
proxmox_virtual_environment_vm.ubuntu_vm: Refreshing state... [id=578]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # proxmox_virtual_environment_vm.ubuntu_vm will be updated in-place
  ~ resource "proxmox_virtual_environment_vm" "ubuntu_vm" {
        id                      = "578"
        name                    = "578-hostpci"
        tags                    = []
        # (23 unchanged attributes hidden)

      - cpu {
          - cores      = 1 -> null
          - flags      = [] -> null
          - hotplugged = 0 -> null
          - numa       = false -> null
          - sockets    = 1 -> null
          - type       = "qemu64" -> null
          - units      = 1024 -> null
        }

      + hostpci {
          + device = "hostpci0"
          + id     = "0000:01:00.0"
          + pcie   = true
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```
(i know there are some issues with `cpu` section, but it is a different story :)
 
<img width="734" alt="Screenshot 2023-09-24 at 8 56 48 PM" src="https://github.com/bpg/terraform-provider-proxmox/assets/627562/f72cef4c-e5d2-4985-a21f-cd26f9475c42">

Wondering, if I'm missing something in the use case?